src/Orleans/Storage/GenericRelationalStorage.cs

```diff
@@ -0,0 +1,210 @@
+        /// <param name="parameterProvider">Adds parameters to the query. Parameter names must match those defined in the query.</param>
+        /// <returns>Affected rows count.</returns>
+        public async Task<int> ExecuteAsync(string query, Action<IDbCommand> parameterProvider)
+        {
+            //If the query is something else that is not acceptable (e.g. an empty string), there will an appropriate database exception.
+            if(query == null)
+            {
+                throw new ArgumentNullException("query");
+            }
+
+            return (await ExecuteAsync(query, parameterProvider, ExecuteReaderAsync, (unit, id) => unit, factory, connectionString).ConfigureAwait(continueOnCapturedContext: false)).Item2;
+        }
+
+
+        protected GenericRelationalStorage(string invariantName, string connectionString)
+        {
+            factory = DbProviderFactories.GetFactory(invariantName);
+            this.connectionString = connectionString;
+            this.invariantName = invariantName;
+        }
+
+
+        private static async Task<Tuple<IEnumerable<TResult>, int>> SelectAsync<TResult>(DbDataReader reader, Func<IDataReader, int, TResult> selector)
+        {
+            var results = new List<TResult>();
+            int resultSetCount = 0;
+            while(reader.HasRows)
+            {
+                while(await reader.ReadAsync().ConfigureAwait(continueOnCapturedContext: false))
+                {
+                    var obj = selector(reader, resultSetCount);
+                    results.Add(obj);
+                }
+
+                await reader.NextResultAsync();
+                ++resultSetCount;
+            }
+
+            return Tuple.Create(results.AsEnumerable(), reader.RecordsAffected);
+        }
+
+
+        private static async Task<Tuple<IEnumerable<TResult>, int>> ExecuteReaderAsync<TResult>(DbCommand command, Func<IDataRecord, int, TResult> selector)
+        {
+            using(var reader = await command.ExecuteReaderAsync().ConfigureAwait(continueOnCapturedContext: false))
+            {
+                return await SelectAsync(reader, selector).ConfigureAwait(false);
+            }
+        }
+
+
+        private static async Task<Tuple<IEnumerable<TResult>, int>> ExecuteAsync<TResult>(
+            string query,
+            Action<DbCommand> parameterProvider,
+            Func<DbCommand, Func<IDataRecord, int, TResult>, Task<Tuple<IEnumerable<TResult>, int>>> executor,
+            Func<IDataRecord, int, TResult> selector,
+            DbProviderFactory factory,
+            string connectionString)
+        {
+            using(var connection = factory.CreateConnection())
+            {
+                connection.ConnectionString = connectionString;
+                await connection.OpenAsync().ConfigureAwait(continueOnCapturedContext: false);
+                using(var command = connection.CreateCommand())
```

Comment:
Here, and see before, I've just split the various stages into separate functions and provided "a hook" via a supplied ``Func`` for the user to give parameters to query that needs to started first in order to create the vendor agnostic command object its parameters. Everything is async and configured without flowing the context.